utilities/include/vec4.hpp

```diff
@@ -0,0 +1,149 @@
+#ifndef VEC4_HPP
+#define VEC4_HPP
+
+#include <ostream>
+#include <cmath>
+#include <xsimd/xsimd.hpp>
+
+#define ALPHA_ 0 //include alpha or not.
+
+#if XSIMD_WITH_AVX512BW || XSIMD_WITH_AVX512DQ || XSIMD_WITH_AVX512F
+    #define ALIGN_WIDTH 64
+#elif XSIMD_WITH_FMA3_AVX2 || XSIMD_WITH_AVX2 || XSIMD_WITH_FMA3_AVX || XSIMD_WITH_AVX
+    #define ALIGN_WIDTH 32
+#elif XSIMD_WITH_SSE4_2 || XSIMD_WITH_SSE4_1 || XSIMD_WITH_SSSE3 || XSIMD_WITH_SSE3 || XSIMD_WITH_SSE2
+    #define ALIGN_WIDTH 16
+#endif
+
+template<class T, class A = xsimd::default_arch>
+class alignas(ALIGN_WIDTH) vec4{
+    public:
+        constexpr vec4() : m_v{0,0,0,0} {}
+        constexpr vec4(T x) : m_v{x,x,x,x} {}
+        constexpr vec4(T x,T y,T z) : m_v{x,y,z,0} {}
+        constexpr vec4(T x,T y,T z,T a) : m_v{x,y,z,a} {}
+        constexpr vec4(xsimd::batch<T,A> x){x.store_aligned(m_v);}
+        
+        //getters
+        constexpr T x() const{
+            return m_v[0];
+        }
+
+        constexpr T y() const{
+            return m_v[1];
+        }
+        
+        constexpr T z() const{
+            return m_v[2];
+        }
+
+        constexpr T a() const{
+            return m_v[3];
+        }
+        
+
+        constexpr T dot(const vec4& v2,const bool alpha=ALPHA_) const{
+            return m_v[0]*v2.m_v[0]+m_v[1]*v2.m_v[1]+m_v[2]*v2.m_v[2]+(m_v[3]*v2.m_v[3])*alpha;
```

Comment:
The vec4 class shouldn't give the user the option to ignore the 4th element. What it _should_ do is act like a proper `vec4` class with all 4 elements. Then, we can inherit from it in a new class `vec3`, and offer the functionalities via inheritance.

As far as the user is concerned, they will use `vec3` when they three-dimensional vectors, and `vec4` four-dimensional vectors, without knowing that `vec3` is just `vec4` with the class throwing out the last element.

Reply:
Vec4 doesn't even require many functions like dot product right? I will create a single file Vec.h which will contain Vec4 and Vec3.

Reply:
Yes, but if we define `dot` in vec3 and later on move to vec4, then we won't have a dot product. For this reason, everything which is common to all types of vectors regardless of their dimensionality should be in `vec4`.

Reply:
Do we need a dot in Vec4 or not?

Reply:
It is best to have it in `vec4` instead of `vec3`.